Projects/Server/Gumps/New/Components/GumpBuilder.cs

```diff
@@ -0,0 +1,698 @@
+            bool background = false, bool scrollbar = false)
+        {
+            WriteStart(Labels.HtmlLocalizedWithArgs);
+            WriteValue(x);
+            WriteValue(y);
+            WriteValue(width);
+            WriteValue(height);
+            WriteValue(background);
+            WriteValue(scrollbar);
+            WriteValue(color);
+            WriteValue(number);
+            WriteValue('@');
+            WriteValue(args);
+            WriteValue('@');
+            WriteEnd();
+        }
+
+        public void AddImage(int x, int y, int gumpId, int hue = 0, ReadOnlySpan<char> @class = default)
+        {
+            WriteStart(Labels.Image);
+            WriteValue(x);
+            WriteValue(y);
+            WriteValue(gumpId);
+
+            if (hue != 0)
+            {
+                Write(HueAttr);
+                WriteValue(hue);
+            }
+
+            if (!@class.IsEmpty)
+            {
+                Write(ClassAttr);
+                WriteValue(@class);
+            }
+
+            WriteEnd();
+        }
+
+        public void AddImageTiledButton(int x, int y, int normalId, int pressedId, int buttonId, GumpButtonType type, int param,
+            int itemId, int hue, int width, int height, int localizedTooltip = -1)
+        {
+            WriteStart(Labels.ImageTileButton);
+            WriteValue(x);
+            WriteValue(y);
+            WriteValue(normalId);
+            WriteValue(pressedId);
+            WriteValue(type, "D");
+            WriteValue(param);
+            WriteValue(buttonId);
+            WriteValue(itemId);
+            WriteValue(hue);
+            WriteValue(width);
+            WriteValue(height);
+            WriteEnd();
+
+            if (localizedTooltip <= 0)
+            {
+                return;
+            }
+
+            WriteStart(Labels.Tooltip);
+            WriteValue(localizedTooltip);
+            WriteEnd();
+        }
+
+        public void AddImageTiled(int x, int y, int width, int height, int gumpId)
+        {
+            WriteStart(Labels.ImageTiled);
+            WriteValue(x);
+            WriteValue(y);
+            WriteValue(width);
+            WriteValue(height);
+            WriteValue(gumpId);
+            WriteEnd();
+        }
+
+        public void AddItem(int x, int y, int itemId, int hue = 0)
+        {
+            WriteStart(hue == 0 ? Labels.Item : Labels.ItemHued);
+            WriteValue(x);
+            WriteValue(y);
+            WriteValue(itemId);
+
+            if (hue != 0)
+            {
+                WriteValue(hue);
+            }
+
+            WriteEnd();
+        }
+
+        public void AddItemProperty(Serial serial)
+        {
+            WriteStart(Labels.ItemProperty);
+            WriteValue(serial.Value);
+            WriteEnd();
+        }
+
+        public void AddLabel(int x, int y, int hue, ReadOnlySpan<char> text)
```

Comment:
For everything that takes text, let's add (in addition to ReadOnlySpan<char>), let's add support for RawInterpolatedStringHandler. The advantage of that is when we do this:

```cs
builder.AddLabel(x, y, hue, $"key_{slotKey}");
```

We are not creating an intermediary string that is allocated and discarded. The `RawInterpolatedStringHandler` uses a rented buffer. See `SpanWriter` for how to use it, it is like the other ones.

Reply:
I figured out the issue. I'll make adjustments.